File: kreuzberg/packages/r/R/plugins.R
# --- OCR Backend Plugins ---

#' Register a custom OCR backend
#'
#' @param name Character string naming the backend.
#' @param callback Callback object for OCR processing.
#' @return Invisible NULL on success.
#' @export
register_ocr_backend <- function(name, callback) {
  stopifnot(is.character(name), length(name) == 1L)
  check_native_result(register_ocr_backend_native(name, callback))
}

#' Unregister a custom OCR backend
#'
#' @param name Character string naming the backend to remove.
#' @return Invisible NULL on success.
#' @export
unregister_ocr_backend <- function(name) {
  stopifnot(is.character(name), length(name) == 1L)
  check_native_result(unregister_ocr_backend_native(name))
}

#' List registered OCR backends
#'
#' @return Character vector of registered backend names.
#' @export
list_ocr_backends <- function() {
  check_native_result(list_ocr_backends_native())
}

#' Clear all registered OCR backends
#'
#' @return Invisible NULL on success.
#' @export
clear_ocr_backends <- function() {
  check_native_result(clear_ocr_backends_native())
}

# --- Post-Processor Plugins ---

#' Register a custom post-processor
#'
#' @param name Character string naming the post-processor.
#' @param callback Callback object for post-processing.
#' @return Invisible NULL on success.
#' @export
register_post_processor <- function(name, callback) {
  stopifnot(is.character(name), length(name) == 1L)
  check_native_result(register_post_processor_native(name, callback))
}

#' Unregister a custom post-processor
#'
#' @param name Character string naming the post-processor to remove.
#' @return Invisible NULL on success.
#' @export
unregister_post_processor <- function(name) {
  stopifnot(is.character(name), length(name) == 1L)
  check_native_result(unregister_post_processor_native(name))
}

#' List registered post-processors
#'
#' @return Character vector of registered post-processor names.
#' @export
list_post_processors <- function() {
  check_native_result(list_post_processors_native())
}

#' Clear all registered post-processors
#'
#' @return Invisible NULL on success.
#' @export
clear_post_processors <- function() {
  check_native_result(clear_post_processors_native())
}

# --- Validator Plugins ---

#' Register a custom validator
#'
#' @param name Character string naming the validator.
#' @param callback Callback object for validation.
#' @return Invisible NULL on success.
#' @export
register_validator <- function(name, callback) {
  stopifnot(is.character(name), length(name) == 1L)
  check_native_result(register_validator_native(name, callback))
}

#' Unregister a custom validator
#'
#' @param name Character string naming the validator to remove.
#' @return Invisible NULL on success.
#' @export
unregister_validator <- function(name) {
  stopifnot(is.character(name), length(name) == 1L)
  check_native_result(unregister_validator_native(name))
}

#' List registered validators
#'
#' @return Character vector of registered validator names.
#' @export
list_validators <- function() {
  check_native_result(list_validators_native())
}

#' Clear all registered validators
#'
#' @return Invisible NULL on success.
#' @export
clear_validators <- function() {
  check_native_result(clear_validators_native())
}

# --- Document Extractor Plugins ---

#' List registered document extractors
#'
#' @return Character vector of registered extractor names.
#' @export
list_document_extractors <- function() {
  check_native_result(list_document_extractors_native())
}

#' Unregister a custom document extractor
#'
#' @param name Character string naming the extractor to remove.
#' @return Invisible NULL on success.
#' @export
unregister_document_extractor <- function(name) {
  stopifnot(is.character(name), length(name) == 1L)
  check_native_result(unregister_document_extractor_native(name))
}

#' Clear all registered document extractors
#'
#' @return Invisible NULL on success.
#' @export
clear_document_extractors <- function() {
  check_native_result(clear_document_extractors_native())
}

File: kreuzberg/packages/r/src/rust/src/plugins.rs
//! Plugin registration FFI wrappers

use crate::error::kreuzberg_error;
use extendr_api::prelude::*;

use kreuzberg::plugins::{
    unregister_validator as kz_unregister_validator,
    clear_validators as kz_clear_validators,
    list_validators as kz_list_validators,
    list_post_processors as kz_list_post_processors,
    list_extractors as kz_list_extractors,
    unregister_extractor as kz_unregister_extractor,
    clear_extractors as kz_clear_extractors,
    unregister_ocr_backend as kz_unregister_ocr_backend,
    list_ocr_backends as kz_list_ocr_backends,
    clear_ocr_backends as kz_clear_ocr_backends,
};

// Post-processor plugins
pub fn register_post_processor_impl(_name: &str, _callback: Robj) -> extendr_api::Result<()> {
    Err(extendr_api::Error::Other("Post-processor registration from R not yet supported".to_string()))
}

pub fn unregister_post_processor_impl(name: &str) -> extendr_api::Result<()> {
    let registry = kreuzberg::get_post_processor_registry();
    registry
        .write()
        .remove(name)
        .map_err(kreuzberg_error)?;
    Ok(())
}

pub fn list_post_processors_impl() -> extendr_api::Result<Strings> {
    let names = kz_list_post_processors().map_err(kreuzberg_error)?;
    Ok(Strings::from_values(names))
}

pub fn clear_post_processors_impl() -> extendr_api::Result<()> {
    let registry = kreuzberg::get_post_processor_registry();
    registry
        .write()
        .shutdown_all()
        .map_err(kreuzberg_error)?;
    Ok(())
}

// Validator plugins
pub fn register_validator_impl(_name: &str, _callback: Robj) -> extendr_api::Result<()> {
    Err(extendr_api::Error::Other("Validator registration from R not yet supported".to_string()))
}

pub fn unregister_validator_impl(name: &str) -> extendr_api::Result<()> {
    kz_unregister_validator(name).map_err(kreuzberg_error)
}

pub fn list_validators_impl() -> extendr_api::Result<Strings> {
    let names = kz_list_validators().map_err(kreuzberg_error)?;
    Ok(Strings::from_values(names))
}

pub fn clear_validators_impl() -> extendr_api::Result<()> {
    kz_clear_validators().map_err(kreuzberg_error)
}

// OCR backend plugins
pub fn register_ocr_backend_impl(_name: &str, _callback: Robj) -> extendr_api::Result<()> {
    Err(extendr_api::Error::Other("OCR backend registration from R not yet supported".to_string()))
}

pub fn unregister_ocr_backend_impl(name: &str) -> extendr_api::Result<()> {
    kz_unregister_ocr_backend(name).map_err(kreuzberg_error)
}

pub fn list_ocr_backends_impl() -> extendr_api::Result<Strings> {
    let names = kz_list_ocr_backends().map_err(kreuzberg_error)?;
    Ok(Strings::from_values(names))
}

pub fn clear_ocr_backends_impl() -> extendr_api::Result<()> {
    kz_clear_ocr_backends().map_err(kreuzberg_error)
}

// Document extractor plugins
pub fn list_document_extractors_impl() -> extendr_api::Result<Strings> {
    let names = kz_list_extractors().map_err(kreuzberg_error)?;
    Ok(Strings::from_values(names))
}

pub fn unregister_document_extractor_impl(name: &str) -> extendr_api::Result<()> {
    kz_unregister_extractor(name).map_err(kreuzberg_error)
}

pub fn clear_document_extractors_impl() -> extendr_api::Result<()> {
    kz_clear_extractors().map_err(kreuzberg_error)
}
